doc/new-syntax.md

```diff
@@ -0,0 +1,337 @@
+# Goal
+
+Simphony code is valid Rust code _(assuming that used types, functions and constants are imported)_.
+
+Writing sum values as `Either<A, B>` instead of `A + B` might seem pointless or verbose.
+
+However, using syntax that is already familiar every Rust developer out there will make it a lot easier to use Simphony.
+
+Rust has everything we need for Simplicity, so let's define a subset of Rust—called Simphony—that translates to Simplicity.
+
+# Types
+
+| Type           | Equivalent to         | Description                  |
+|----------------|-----------------------|------------------------------|
+| `()`           |                       | Unit type                    |
+| `(A, B)`       |                       | Product of types `A` and `B` |
+| `Either<A, B>` |                       | Sum of types `A` and `B`     |
+| `Option<A>`    | `Either<(), A>`       | Option of type `A`           |
+| `u1`           | `Either<(), ()>`      | Boolean                      |
+| `u2`           | `(u1, u1)`            | 2-bit unsigned integer       |
+| `u4`           | `(u2, u2)`            | 4-bit unsigned integer       |
+| `u8`           | `(u4, u4)`            | 8-bit unsigned integer       |
+| `u16`          | `(u8, u8)`            | 16-bit unsigned integer      |
+| `u32`          | `(u16, u16)`          | 32-bit unsigned integer      |
+| `u64`          | `(u32, u32)`          | 64-bit unsigned integer      |
+| `u128`         | `(u64, u64)`          | 128-bit unsigned integer     |
+| `u256`         | `(u128, u128)`        | 256-bit unsigned integer     |
+| `[A; 1]`       | `Option<A>`           | List of length ≤ 1           |
+| `[A; N + 1]`   | `Option<(A, [A; N])>` | List of length ≤ N + 1       |
+
+The Boolean type is essentially a macro for a sum type of two units.
+
+The unsigned integer types are essentially macros for product types.
+
+The list type `[A; N]` can hold up to `N` values of type `A` where `N` is a positive integer.
+
+We need to know the list size at commit time, so `N` will be somewhat large. There will likely be parts of the list that are unpopulated at redeem time. The definition of `[A; N]` as nested option-pairs keeps the onchain footprint small, since the unpopulated parts don't contribute extra bits to the onchain encoding.
```

Comment:
Honestly I'd be tempted to use the syntax `List<A, N>` rather than `[A; N]` to emphasize that this is something weird and not very array-like. In particular it has linear access time.